File: packages/nest-crud-request/src/lib/query-builder.ts
import { RelationBuilder } from './relation-builder';
import { QueryBuilderOptions, OrderDirectionEnum, AggregateSpec, RelationObjectValue } from './types';
import { deepMerge } from './utils';
import { WhereBuilder, WhereBuilderCondition } from './where-builder';


export class QueryBuilder {

    private options: QueryBuilderOptions = {};

    private whereBuilder: WhereBuilder = new WhereBuilder();

    private havingBuilder: WhereBuilder = new WhereBuilder();

    private relationBuilder: RelationBuilder = new RelationBuilder();

    constructor(options?: QueryBuilderOptions) {
        if (options) {
            this.setOptions(options);
        }
    }

    setOptions(options: QueryBuilderOptions): this {
        this.options = options;
        this.whereBuilder = new WhereBuilder(options.where);
        this.havingBuilder = new WhereBuilder(options.having);
        this.relationBuilder = new RelationBuilder(options.relations);
        return this;
    }

    mergeOptions(options: QueryBuilderOptions, deep = false): this {
        let updatedOptions = {};
        if (deep) {
            updatedOptions = deepMerge(this.options, options);
        } else {
            updatedOptions = {
                ...this.options,
                ...options,
            };
        }
        this.setOptions(updatedOptions);
        return this;
    }

    addSelect(fields: string | string[]): this {
        // Ensure select is always an array
        if (!this.options.select || typeof this.options.select === 'string') {
            this.options.select = [];
        }
        if (Array.isArray(fields)) {
            this.options.select.push(...fields);
        } else {
            this.options.select.push(fields);
        }
        return this;
    }

    removeSelect(fields: string | string[]): this {
        // Ensure select is an array before filtering
        if (this.options.select && Array.isArray(this.options.select)) {
            if (Array.isArray(fields)) {
                this.options.select = this.options.select.filter(field => !fields.includes(field));
            } else {
                this.options.select = this.options.select.filter(field => field !== fields);
            }
        }
        return this;
    }

    addRelation(
        relation: string,
        select?: string[] | RelationObjectValue,
        where?: Record<string, any>,
        joinType?: 'left' | 'inner',
    ): this {
        this.relationBuilder.add(relation, select as any, where, joinType);
        return this;
    }

    removeRelation(relation: string): this {
        this.relationBuilder.remove(relation);
        return this;
    }

    where(...args: WhereBuilderCondition): this {
        this.whereBuilder.where(...args);
        return this;
    }

    andWhere(...args: WhereBuilderCondition): this {
        this.whereBuilder.andWhere(...args);
        return this;
    }

    orWhere(...args: WhereBuilderCondition): this {
        this.whereBuilder.orWhere(...args);
        return this;
    }

    addOrder(orderBy: string, order: OrderDirectionEnum): this {
        if (!this.options.order) {
            this.options.order = {};
        }
        this.options.order[orderBy] = order;
        return this;
    }

    /**
     * Add a per-row aggregate over a relation, e.g.
     * `.addAggregate({ fn: 'count', field: 'posts.id', as: 'postCount' })`.
     * The alias can then be used with `.having(...)` and `.addOrder(...)`.
     */
    addAggregate(aggregate: AggregateSpec): this {
        if (!Array.isArray(this.options.aggregates)) {
            this.options.aggregates = [];
        }
        this.options.aggregates.push(aggregate);
        return this;
    }

    removeAggregate(as: string): this {
        if (Array.isArray(this.options.aggregates)) {
            this.options.aggregates = this.options.aggregates.filter((a) => a.as !== as);
        }
        return this;
    }

    /** Filter on an aggregate alias — same operator syntax as `where`. */
    having(...args: WhereBuilderCondition): this {
        this.havingBuilder.where(...args);
        return this;
    }

    andHaving(...args: WhereBuilderCondition): this {
        this.havingBuilder.andWhere(...args);
        return this;
    }

    orHaving(...args: WhereBuilderCondition): this {
        this.havingBuilder.orWhere(...args);
        return this;
    }

    removeOrder(orderBy: string): this {
        if (this.options.order) {
            delete this.options.order[orderBy];
        }
        return this;
    }

    setSkip(skip: number): this {
        this.options.skip = skip;
        return this;
    }

    setTake(take: number): this {
        this.options.take = take;
        return this;
    }

    setWithDeleted(withDeleted: boolean): this {
        this.options.withDeleted = withDeleted;
        return this;
    }

    setOnlyDeleted(onlyDeleted: boolean): this {
        this.options.onlyDeleted = onlyDeleted;
        return this;
    }

    set(key: string, value: any): this {
        this.options[key] = value;
        return this;
    }
    
    toObject(constrainToNestedObject = false) {
        const options = {
            ...this.options,
        };

        // Convert where conditions to JSON string
        if (this.whereBuilder.hasConditions()) {
            if (constrainToNestedObject) {
                options.where = this.whereBuilder.toObject();
            } else {
                options.where = JSON.stringify(this.whereBuilder.toObject());
            }
        } else {
            delete options.where;
        }

        // Convert relations to JSON string
        if (this.relationBuilder.hasRelations()) {
            if (constrainToNestedObject) {
                options.relations = this.relationBuilder.toObject();
            } else {
                options.relations = JSON.stringify(this.relationBuilder.toObject());
            }
        } else {
            delete options.relations;
        }

        // Convert order to a JSON string unless emitting nested objects
        if (options.order && Object.keys(options.order).length > 0) {
            if (!constrainToNestedObject) {
                options.order = JSON.stringify(options.order);
            }
        } else {
            delete options.order;
        }

        // Convert select to a JSON string unless emitting nested objects
        if (options.select && options.select.length > 0) {
            if (!constrainToNestedObject) {
                options.select = JSON.stringify(options.select);
            }
        } else {
            delete options.select;
        }

        // Serialise aggregates (JSON string unless emitting nested objects)
        if (Array.isArray(this.options.aggregates) && this.options.aggregates.length > 0) {
            options.aggregates = constrainToNestedObject
                ? this.options.aggregates
                : JSON.stringify(this.options.aggregates);
        } else {
            delete options.aggregates;
        }

        // Convert having conditions like where (the builder is the source of truth)
        if (this.havingBuilder.hasConditions()) {
            options.having = constrainToNestedObject
                ? this.havingBuilder.toObject()
                : JSON.stringify(this.havingBuilder.toObject());
        } else {
            delete options.having;
        }

        return options;
    }

    toJson() {
        const obj = this.toObject(true);
        return JSON.stringify(obj);
    }

}

File: clients/flutter/nest_crud_request/lib/src/query_builder.dart
import 'dart:convert';

import 'operators.dart';
import 'relation_builder.dart';
import 'where_builder.dart';

/// Fluent builder for `@ackplus/nest-crud` list-query parameters — the Dart twin
/// of the JS `QueryBuilder` in `@ackplus/nest-crud-request`. It produces the exact
/// same query strings, so it talks to your existing API with no server changes.
///
/// ```dart
/// final params = (QueryBuilder()
///       ..where('status', 'active')
///       ..whereOp('age', WhereOperator.gte, 18)
///       ..addRelation('posts', select: ['id', 'title'])
///       ..addAggregate(fn: AggregateFn.count, field: 'posts.id', as: 'postCount')
///       ..having('postCount', WhereOperator.gt, 5)
///       ..addOrder('createdAt', OrderDirection.desc)
///       ..setTake(20))
///     .toQueryParameters();
///
/// final res = await dio.get('/users', queryParameters: params);
/// ```
class QueryBuilder {
  Map<String, dynamic> _options = {};
  WhereBuilder _where = WhereBuilder();
  WhereBuilder _having = WhereBuilder();
  RelationBuilder _relations = RelationBuilder();

  QueryBuilder([Map<String, dynamic>? options]) {
    if (options != null) setOptions(options);
  }

  QueryBuilder setOptions(Map<String, dynamic> options) {
    _options = Map<String, dynamic>.from(options);
    _where = WhereBuilder(options['where']);
    _having = WhereBuilder(options['having']);
    _relations = RelationBuilder(options['relations']);
    return this;
  }

  QueryBuilder mergeOptions(Map<String, dynamic> options, {bool deep = false}) {
    return setOptions(deep ? _deepMerge(_options, options) : {..._options, ...options});
  }

  // ---- select ----
  QueryBuilder addSelect(dynamic fields) {
    final current = _options['select'] is List ? List<String>.from(_options['select'] as List) : <String>[];
    if (fields is List) {
      current.addAll(fields.cast<String>());
    } else {
      current.add(fields as String);
    }
    _options['select'] = current;
    return this;
  }

  QueryBuilder removeSelect(dynamic fields) {
    if (_options['select'] is List) {
      final remove = fields is List ? fields.cast<String>() : [fields as String];
      _options['select'] = (_options['select'] as List).where((f) => !remove.contains(f)).toList();
    }
    return this;
  }

  // ---- relations ----
  QueryBuilder addRelation(String relation,
      {List<String>? select, Map<String, dynamic>? where, String? joinType}) {
    _relations.add(relation, select: select, where: where, joinType: joinType);
    return this;
  }

  QueryBuilder removeRelation(String relation) {
    _relations.remove(relation);
    return this;
  }

  // ---- where (delegate to WhereBuilder) ----
  QueryBuilder where(String field, dynamic value) { _where.where(field, value); return this; }
  QueryBuilder whereOp(String field, WhereOperator op, dynamic value) { _where.whereOp(field, op, value); return this; }
  QueryBuilder whereRaw(Map<String, dynamic> condition) { _where.whereRaw(condition); return this; }
  QueryBuilder whereGroup(void Function(WhereBuilder) build) { _where.whereGroup(build); return this; }
  QueryBuilder andWhere(String field, dynamic value) { _where.andWhere(field, value); return this; }
  QueryBuilder andWhereOp(String field, WhereOperator op, dynamic value) { _where.andWhereOp(field, op, value); return this; }
  QueryBuilder andWhereGroup(void Function(WhereBuilder) build) { _where.andWhereGroup(build); return this; }
  QueryBuilder orWhere(String field, dynamic value) { _where.orWhere(field, value); return this; }
  QueryBuilder orWhereOp(String field, WhereOperator op, dynamic value) { _where.orWhereOp(field, op, value); return this; }
  QueryBuilder orWhereGroup(void Function(WhereBuilder) build) { _where.orWhereGroup(build); return this; }

  // ---- order ----
  QueryBuilder addOrder(String field, OrderDirection direction) {
    final order = _options['order'] is Map ? Map<String, dynamic>.from(_options['order'] as Map) : <String, dynamic>{};
    order[field] = direction.value;
    _options['order'] = order;
    return this;
  }

  QueryBuilder removeOrder(String field) {
    if (_options['order'] is Map) (_options['order'] as Map).remove(field);
    return this;
  }

  // ---- aggregates ----
  QueryBuilder addAggregate({
    required AggregateFn fn,
    required String field,
    required String as,
    bool? distinct,
    Map<String, dynamic>? where,
  }) {
    final list = _options['aggregates'] is List ? List<dynamic>.from(_options['aggregates'] as List) : <dynamic>[];
    final spec = <String, dynamic>{'fn': fn.value, 'field': field, 'as': as};
    if (distinct != null) spec['distinct'] = distinct;
    if (where != null) spec['where'] = where;
    list.add(spec);
    _options['aggregates'] = list;
    return this;
  }

  QueryBuilder removeAggregate(String as) {
    if (_options['aggregates'] is List) {
      _options['aggregates'] = (_options['aggregates'] as List).where((a) => a is Map && a['as'] != as).toList();
    }
    return this;
  }

  // ---- having (delegate) ----
  QueryBuilder having(String field, dynamic value) { _having.where(field, value); return this; }
  QueryBuilder havingOp(String field, WhereOperator op, dynamic value) { _having.whereOp(field, op, value); return this; }
  QueryBuilder andHaving(String field, dynamic value) { _having.andWhere(field, value); return this; }
  QueryBuilder orHaving(String field, dynamic value) { _having.orWhere(field, value); return this; }

  // ---- pagination / flags / custom ----
  QueryBuilder setSkip(int skip) { _options['skip'] = skip; return this; }
  QueryBuilder setTake(int take) { _options['take'] = take; return this; }
  QueryBuilder setWithDeleted(bool withDeleted) { _options['withDeleted'] = withDeleted; return this; }
  QueryBuilder setOnlyDeleted(bool onlyDeleted) { _options['onlyDeleted'] = onlyDeleted; return this; }
  QueryBuilder set(String key, dynamic value) { _options[key] = value; return this; }

  // ---- output ----

  /// Build the params. By default complex fields (`where`, `relations`, `order`,
  /// `select`, `aggregates`, `having`) are JSON-stringified — the HTTP shape the
  /// server parses. Pass [nested] = true to keep them as native maps/lists.
  Map<String, dynamic> toObject({bool nested = false}) {
    final out = Map<String, dynamic>.from(_options);

    void put(String key, bool present, dynamic native) {
      if (present) {
        out[key] = nested ? native : jsonEncode(native);
      } else {
        out.remove(key);
      }
    }

    put('where', _where.hasConditions(), _where.toObject());
    put('relations', _relations.hasRelations(), _relations.toObject());
    put('having', _having.hasConditions(), _having.toObject());

    final order = _options['order'];
    put('order', order is Map && order.isNotEmpty, order);
    final select = _options['select'];
    put('select', select is List && select.isNotEmpty, select);
    final aggregates = _options['aggregates'];
    put('aggregates', aggregates is List && aggregates.isNotEmpty, aggregates);

    return out;
  }

  /// Params ready for an HTTP client (`dio` / `http`) — every value as a String.
  Map<String, String> toQueryParameters() {
    return toObject().map((key, value) => MapEntry(key, value is String ? value : '$value'));
  }

  /// The whole query as a single JSON string (native/nested form).
  String toJson() => jsonEncode(toObject(nested: true));

  static Map<String, dynamic> _deepMerge(Map<String, dynamic> target, Map<String, dynamic> source) {
    final result = Map<String, dynamic>.from(target);
    source.forEach((key, value) {
      if (value is Map && result[key] is Map) {
        result[key] = _deepMerge(Map<String, dynamic>.from(result[key] as Map), Map<String, dynamic>.from(value));
      } else {
        result[key] = value;
      }
    });
    return result;
  }
}
